In [101]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

#  Import and read the charity_data.csv.
import pandas as pd
application_df = pd.read_csv("https://static.bc-edx.com/data/dl-1-2/m21/lms/starter/charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [102]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.
application_df.drop(['EIN','NAME'],axis=1,inplace=True)

In [103]:
# Determine the number of unique values in each column.
unique_value_count = application_df.nunique()
print(unique_value_count)

APPLICATION_TYPE            17
AFFILIATION                  6
CLASSIFICATION              71
USE_CASE                     5
ORGANIZATION                 4
STATUS                       2
INCOME_AMT                   9
SPECIAL_CONSIDERATIONS       2
ASK_AMT                   8747
IS_SUCCESSFUL                2
dtype: int64


In [104]:
# Look at APPLICATION_TYPE value counts for binning
value_count_application_type = application_df['APPLICATION_TYPE'].value_counts()
print(value_count_application_type)

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64


In [105]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
cutoff_value = 528
application_types_to_replace = list(value_count_application_type[value_count_application_type< cutoff_value].index)
#print(application_types_to_replace)
for app in application_types_to_replace:
    application_df['APPLICATION_TYPE']=application_df['APPLICATION_TYPE'].replace(app,"Other")


In [106]:

application_df['APPLICATION_TYPE'].value_counts()

T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      276
Name: APPLICATION_TYPE, dtype: int64

In [107]:
# Look at CLASSIFICATION value counts for binning
value_count_classification = application_df['CLASSIFICATION'].value_counts()
print(value_count_classification)

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: CLASSIFICATION, Length: 71, dtype: int64


In [108]:
# Filter the value counts to show only counts greater than 1
classification_counts = application_df['CLASSIFICATION'].value_counts()


classification_counts_gt1 = classification_counts[classification_counts > 1]

print(classification_counts_gt1)

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
C7120       18
C1500       16
C1800       15
C6000       15
C1250       14
C8200       11
C1238       10
C1278       10
C1235        9
C1237        9
C7210        7
C2400        6
C1720        6
C4100        6
C1257        5
C1600        5
C1260        3
C2710        3
C0           3
C3200        2
C1234        2
C1246        2
C1267        2
C1256        2
Name: CLASSIFICATION, dtype: int64


In [109]:
# Choose a cutoff value and create a list of classifications to be replaced
# use the variable name `classifications_to_replace`
cutoff_value_classification = 1883
classifications_to_replace = list(classification_counts[classification_counts< cutoff_value_classification].index)
#print(application_types_to_replace)
for app in classifications_to_replace:
    application_df['CLASSIFICATION']=application_df['CLASSIFICATION'].replace(app,"Other")

In [110]:
 application_df['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
Other     2261
C3000     1918
C2100     1883
Name: CLASSIFICATION, dtype: int64

In [111]:
# Convert categorical data to numeric with `pd.get_dummies`
categorical_columns = ['APPLICATION_TYPE', 'CLASSIFICATION','AFFILIATION','USE_CASE','ORGANIZATION','SPECIAL_CONSIDERATIONS','INCOME_AMT']

application_df = pd.get_dummies(application_df,columns=categorical_columns)

In [112]:
application_df

STATUS   ASK_AMT  IS_SUCCESSFUL  APPLICATION_TYPE_Other  \
0           1      5000              1                       0   
1           1    108590              1                       0   
2           1      5000              0                       0   
3           1      6692              1                       0   
4           1    142590              1                       0   
...       ...       ...            ...                     ...   
34294       1      5000              0                       0   
34295       1      5000              0                       0   
34296       1      5000              0                       0   
34297       1      5000              1                       0   
34298       1  36500179              0                       0   

       APPLICATION_TYPE_T10  APPLICATION_TYPE_T19  APPLICATION_TYPE_T3  \
0                         1                     0                    0   
1                         0                     0                    1   
2                         0                     0                    0   
3                         0                     0                    1   
4                         0                     0                    1   
...                     ...                   ...                  ...   
34294                     0                     0                    0   
34295                     0                     0                    0   
34296                     0                     0                    1   
34297                     0                     0                    0   
34298                     0                     0                    1   

       APPLICATION_TYPE_T4  APPLICATION_TYPE_T5  APPLICATION_TYPE_T6  ...  \
0                        0                    0                    0  ...   
1                        0                    0                    0  ...   
2                        0                    1                    0  ...   
3                        0                    0                    0  ...   
4                        0                    0                    0  ...   
...                    ...                  ...                  ...  ...   
34294                    1                    0                    0  ...   
34295                    1                    0                    0  ...   
34296                    0                    0                    0  ...   
34297                    0                    1                    0  ...   
34298                    0                    0                    0  ...   

       SPECIAL_CONSIDERATIONS_Y  INCOME_AMT_0  INCOME_AMT_1-9999  \
0                             0             1                  0   
1                             0             0                  1   
2                             0             1                  0   
3                             0             0                  0   
4                             0             0                  0   
...                         ...           ...                ...   
34294                         0             1                  0   
34295                         0             1                  0   
34296                         0             1                  0   
34297                         0             1                  0   
34298                         0             0                  0   

       INCOME_AMT_10000-24999  INCOME_AMT_100000-499999  INCOME_AMT_10M-50M  \
0                           0                         0                   0   
1                           0                         0                   0   
2                           0                         0                   0   
3                           1                         0                   0   
4                           0                         1                   0   
...                       ...                       ...                 ...   
34294                       0 

In [113]:
# Split our preprocessed data into our features and target arrays
target_column = 'IS_SUCCESSFUL'
X = application_df.drop(columns= [target_column])
y = application_df[target_column]
# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42, stratify=y)

In [114]:
# Create a StandardScaler instances
scaler = StandardScaler()



In [115]:
X_scaler = scaler.fit(X_train)

In [116]:
# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

In [117]:
from keras.api._v2.keras import activations

# Keras
from tensorflow import keras

from tensorflow.keras.utils import to_categorical
from tensorflow.keras.layers import Dense
from tensorflow.keras.datasets import mnist
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.


nn = tf.keras.models.Sequential()




# First hidden layer
nn.add(Dense(80,activation='relu', input_dim = X_train_scaled.shape[1]))

# Second hidden layer
nn.add(Dense(30,activation = 'relu'))
# Third hidden layer
nn.add(Dense(10,activation = 'relu'))

# Output layer
nn.add(Dense(1, activation = 'softmax'))

# Check the structure of the model
nn.summary()

Model: "sequential_5"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_3 (Dense)             (None, 80)                3520      
                                                                 
 dense_4 (Dense)             (None, 30)                2430      
                                                                 
 dense_5 (Dense)             (None, 10)                310       
                                                                 
 dense_6 (Dense)             (None, 1)                 11        
                                                                 
Total params: 6,271
Trainable params: 6,271
Non-trainable params: 0
_________________________________________________________________


In [119]:
# Compile the model
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [120]:
# Train the model
nn.fit(
    X_train,
    y_train,
    epochs=100,
    shuffle=True,
    verbose=2
)

Epoch 1/100
804/804 - 3s - loss: 138.3506 - accuracy: 0.5324 - 3s/epoch - 4ms/step
Epoch 2/100
804/804 - 1s - loss: 0.6911 - accuracy: 0.5324 - 1s/epoch - 2ms/step
Epoch 3/100
804/804 - 1s - loss: 0.6911 - accuracy: 0.5324 - 1s/epoch - 2ms/step
Epoch 4/100
804/804 - 1s - loss: 0.6911 - accuracy: 0.5324 - 1s/epoch - 2ms/step
Epoch 5/100
804/804 - 1s - loss: 0.6911 - accuracy: 0.5324 - 1s/epoch - 2ms/step
Epoch 6/100
804/804 - 1s - loss: 0.6911 - accuracy: 0.5324 - 1s/epoch - 2ms/step
Epoch 7/100
804/804 - 1s - loss: 0.6911 - accuracy: 0.5324 - 1s/epoch - 2ms/step
Epoch 8/100
804/804 - 2s - loss: 0.6911 - accuracy: 0.5324 - 2s/epoch - 3ms/step
Epoch 9/100
804/804 - 2s - loss: 0.6911 - accuracy: 0.5324 - 2s/epoch - 2ms/step
Epoch 10/100
804/804 - 1s - loss: 0.6911 - accuracy: 0.5324 - 1s/epoch - 2ms/step
Epoch 11/100
804/804 - 1s - loss: 0.6911 - accuracy: 0.5324 - 1s/epoch - 2ms/step
Epoch 12/100
804/804 - 1s - loss: 0.6911 - accuracy: 0.5324 - 1s/epoch - 2ms/step
Epoch 13/100
804/804 - 

In [121]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - loss: 0.6931 - accuracy: 0.5324 - 779ms/epoch - 3ms/step
Loss: 0.6931056380271912, Accuracy: 0.5323615074157715


In [122]:
# Export our model to HDF5 file
nn.save("AlphabetSoupCharity_Optimization.h5")

#### Report on Neural network Model
----
##### Purpose:

The Purpose of the analysis is to develop a tool that can help to select the applicants for fundng with the best chance of success in the ventures for Alphabet Soup.

##### Result Analysis
Based on the provided training results, it appears that the model achieved an accuracy of approximately 53.24% on the training data. The loss value at the end of training is approximately 0.6931.

##### breakdown of the training results:

Loss: 0.6931056380271912
Accuracy: 0.5323615074157715 (53.24%)
Explanation:

The "loss" value represents the value of the loss function (in this case, binary cross-entropy) at the end of the training process. Lower values of the loss function indicate better model performance.
The "accuracy" value is the accuracy of the model on the training data, expressed as a percentage. It indicates the proportion of correct predictions made by the model out of all training data samples.

After training, it's recommended to evaluate the model on the test dataset to get a more accurate representation of its performance. This will provide a better estimate of how well the model will perform on new, unseen data. Additionally, further model optimization and tuning may be necessary to achieve better accuracy and generalization.
- The type of APPLICATION is one of the following; T3, T4, T5, T6, T7, T8, T10, and T19
- The application has the following CLASSIFICATION; C1000, C2000, C3000, C1200, and C2100.

#### Recommendation
to increase the accuracyy A good model to recommend is the Random Forest model because Random Forest are good for classification problems.  